### PR TITLE
Raise errors on invalid axis limits

### LINKED ISSUE #7460: Raise error if argument to xlim is invalid, e.g., nan
As of 2.0b4, passing `nan` as a limit to `xlim` seems to reset *both* limits to `(-0.001, 0.001)` (try e.g. with `plt.plot([0, 1]); plt.xlim(right=np.nan)`).  I think this should either error, or, more usefully, be interpreted as None (i.e. leave that limit as is).  (For extra points, we can also bikeshed what passing `inf` to `xlim` should mean.)

The main use case would be when the `xlim` is read from some float context (e.g. from an array) where you can easily store `nan` but not `None`.

### PR DESCRIPTION
**Pull request by alixxxin**
Raise error when xlim is set to nan or inf, same with both set_xlim, set_ylim, set_xlim3d, set_ylim3d, set_zlim3d. 
Tests are included too

Fixes #7460

### CODE REVIEW

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -602,6 +602,10 @@ def set_xlim3d(self, left=None, right=None, emit=True, auto=False, **kw):
         if right is None and cbook.iterable(left):
             left, right = left
 
+        if ((np.isscalar(left)) and not (np.isfinite(left))) \
```
Comment (tacaswell) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Don't need the trailing `\`, enclosing in `()` does the job.

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -2873,6 +2873,11 @@ def set_xlim(self, left=None, right=None, emit=True, auto=False, **kw):
             left, right = left
 
         self._process_unit_info(xdata=(left, right))
+
+        if ((np.isscalar(left)) and not (np.isfinite(left))) \
```
Comment (tacaswell) [inline lib/matplotlib/axes/_base.py]: This should probably go _after_ converting the units.

**Comment by tacaswell** [comment]
github-foo linkage: #7460

**Comment by tacaswell** [comment]
#7744 fixed the bug.  Can this PR be re-based and sort out if the tests should be expanded?

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 1 above)
Comment (alixxxin) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Done

File: lib/matplotlib/axes/_base.py (hunk 2 above)
Comment (alixxxin) [inline lib/matplotlib/axes/_base.py]: Done

**Comment by dstansby** [comment]
I've just changed the title to make it more descriptive; it's much more useful to have a short description in the title than just the bug number.

It also looks like there's some PEP8 errors - you can see them by clicking on the travis link below.

**Comment by QuLogic** [comment]
FYI, those last three commits have some odd emails for the author/committer.

**Comment by tacaswell** [comment]
Do we still need this PR now that #7744 has been merged?  I think the tests are more through from this one?

**Comment by QuLogic** [comment]
I think #8022 might cover even more tests.

**Comment by QuLogic** [comment]
This PR is covered by #7744 and #8022; sorry about that @alixxxin. We hope that doesn't deter you from contributing in the future.

**Pull request closed without merging**